File: spreader3.py
from itertools import cycle

def remove_and_rotate(l, item):
    offset = l.index(item)
    return l[offset+1:] + l[:offset]

def spreader_generator(blockpool, spread):
    blockpool_iter = iter(blockpool)
    feeders = [feeder_generator(blockpool_iter) for _ in range(spread)]
    feeders_iter = cycle(feeders)
    while len(feeders):
        feeder = feeders_iter.next()
        try:
            yield feeder.next()
        except StopIteration:
            feeders = remove_and_rotate(feeders, feeder)
            feeders_iter = cycle(feeders)

def feeder_generator(blockpool):
 for block in blockpool:
     for v in block:
         yield v

File: spreader3.php
<?
class Spreader3 {
	function __construct($blockpool, $spread) {
		reset($blockpool);
		$this->blockpool = $blockpool;
		$this->spread = min(count($blockpool), $spread);
		$this->feeders = array();
		foreach (range(1, $this->spread) as $_) {
			list($__, $feeder) = each($this->blockpool);
			reset($feeder);
			$this->feeders[] = $feeder;
		}
		$this->current_index = 0;
	}
	function next() {
		$found = FALSE;
		do {
			if ($this->spread < 1) {
				throw new StopIteration();
			}
			list($key, $next) = each($this->feeders[$this->current_index]);
			if ($key !== NULL) {
				$found = TRUE;
			} else {
				list($key, $this->feeders[$this->current_index]) = each($this->blockpool);
				if ($key === NULL) {
					array_splice($this->feeders, $this->current_index, 1);
					$this->spread--;
					if ($this->spread > 0) {
						$this->current_index %= $this->spread;
					}
					continue;
				}
				continue;
			}
			$this->current_index = ($this->current_index + 1) % $this->spread;
		} while (!$found);
		return $next;
	}
}
